site/_posts/2017-10-15-fast-python-serialization-with-ray-and-arrow.markdown

```diff
@@ -0,0 +1,198 @@
+are impressive.
+
+<div align="center">
+<img src="{{ site.base-url }}/assets/fast_python_serialization_with_ray_and_arrow/speedups0.png" width="365" height="255">
+<img src="{{ site.base-url }}/assets/fast_python_serialization_with_ray_and_arrow/speedups1.png" width="365" height="255">
+</div>
+
+Note that the biggest wins are with deserialization. The speedups here are
+multiple orders of magnitude and get better as the numpy arrays get larger
+(thanks to design goals 1, 3, and 4). Making **deserialization** fast is
+important for two reasons. First, an object may be serialized once and then
+deserialized many times (e.g., an object that is broadcast to all workers).
+Second, a common pattern is for many objects to be serialized in parallel and
+then aggregated and deserialized one at a time on a single worker making
+deserialization the bottleneck.
+
+**Without numpy arrays:** When using regular Python objects, for which we
+cannot take advantage of shared memory, the results are comparable to pickle.
+
+<div align="center">
+<img src="{{ site.base-url }}/assets/fast_python_serialization_with_ray_and_arrow/speedups2.png" width="365" height="255">
+<img src="{{ site.base-url }}/assets/fast_python_serialization_with_ray_and_arrow/speedups3.png" width="365" height="255">
+</div>
+
+These are just a few examples of interesting Python objects. The most important
+case is the case where numpy arrays are nested within other objects. Note that
+our serialization library works with very general Python types including custom
+Python classes and deeply nested objects.
+
+## Data Representation
+
+We use Apache Arrow as the underlying language-independent data layout. Objects
+are stored in two parts: a **schema** and a **data blob**. At a high level, the
+data blob is roughly a flattened concatenation of all of the data values
+recursively contained in the object, and the schema defines the types and
+nesting structure of the data blob.
+
+Python sequences (e.g., dictionaries, lists, tuples, sets) are encoded as
+[UnionArrays][8] of other types (e.g., bools, ints, strings, bytes, floats,
+doubles, date64s, tensors (i.e., numpy arrays), lists, tuples, dicts and sets).
+Nested sequences are encoded using [ListArrays][9]. All tensors are collected
+and appended to the end of the serialized object, and the UnionArray contains
+references to these tensors.
+
+To give a concrete example, consider the following object.
+
+```python
+[(1, 2), 'hello', 3, 4, np.array([5.0, 6.0])]
+```
+
+It would be represented in Arrow with the following structure.
+
+```
+UnionArray(type_ids=[tuple, string, int, int, ndarray],
+           tuples=ListArray(offsets=[0, 2],
+                            UnionArray(type_ids=[int, int],
+                                       ints=[1, 2])),
+           strings=[“hello”],
+           ints=[3, 4],
+           ndarrays=[<offset of numpy array>])
+```
+
+Arrow encodes the schema as a flatbuffers object. **Using only the schema, we
```

Comment:
"as a flatbuffers object" might be more precise to say "Arrow uses Flatbuffers to encode serialized schemas"